File: perl/mysql-fetch-many-rows-with-low-memory.pl
#!/usr/bin/perl -w
use strict;
use DBI;
use Dumpvalue;
use Time::HiRes qw/usleep/;

local $| = 1;
my $database = "apviaapps_development";
my $hostname = "localhost";
my $port     = "3306";
my $user     = "apviaapps_user";
my $passwd   = "";
my $dsn      = "DBI:mysql:database=$database;host=$hostname;port=$port";
my $option   = {
    RaiseError => 1,
};
my $dbh = DBI->connect( $dsn, $user, $passwd );

# 対象レコードの ID だけを取得
my $sth = $dbh->prepare( "SELECT id FROM action_logs where app_id = 'bunshin' and event_id = 'q_answered'" );
$sth->execute;

my $sth2 = $dbh->prepare( "SELECT * FROM action_logs where id = ?" );
while ( my $res = $sth->fetchrow_arrayref ){
    my $id = $res->[0];
    # 各レコードの詳細情報を毎回取得 (メモリ節約)
    $sth2->execute( $id );
    my $res2 = $sth2->fetchrow_arrayref;
    print $res2->[0], "\n";
    usleep 200_000;
}
print "\n";

File: ruby/mysql-fetch-many-rows-with-low-memory.rb
#!/usr/bin/local/ruby -w
require 'rubygems'
gem 'mysql'
require 'mysql'

begin
    dbh = Mysql.real_connect "localhost", "apviaapps_user", "", "apviaapps_development"

    sth = dbh.prepare "SELECT id from action_logs where app_id = ? and event_id = ?"
    sth.execute 'bunshin', 'q_answered'

    sth2 = dbh.prepare "SELECT * from action_logs where id = ?"
    sth.each do |r|
        sth2.execute r[0]
        sth2.each do |r2|
            p r2[0]
        end
        sth2.free_result
    end
rescue Mysql::Error => e
    puts "Error code: #{e.errno}"
    puts "Error message: #{e.error}"
    puts "Error SQLSTATE: #{e.sqlstate}" if e.respond_to? "sqlstate"
ensure
    dbh.close if dbh
end
